models.py

```diff
@@ -0,0 +1,248 @@
+from payments import db
+# db will be our connect to SQLAlchemy ORM
+
+######################################################################
+# Custom Exceptions
+######################################################################
+class DataValidationError(ValueError):
+    pass
+
+######################################################################
+# Models
+######################################################################
+
+class Payment(db.Model):
+    __tablename__ = 'payment'
+    id = db.Column(db.Integer, primary_key=True)
+    nickname = db.Column(db.String(20))
+    user_id = db.Column(db.Integer)
+    payment_type = db.Column(db.String(10))
+    is_default = db.Column(db.Boolean)
+    is_removed = db.Column(db.Boolean)
+    charge_history = db.Column(db.Float)
+
+    detail_id = db.Column(db.Integer, db.ForeignKey('detail.id'))
+    details = db.relationship('Detail',
+        backref=db.backref('payment', lazy='joined'))
+
+    def self_url(self):
+        return url_for('get_payments', id=self.id, _external=True)
+
+    def serialize(self):
+        return {
+                    'payment_id' : self.payment_id,
+                    'user_id' : self.user_id,
+                    'nickname' : self.nickname,
+                    'payment_type' : self.payment_type,
+                    'is_default' : self.is_default,
+                    'charge_history' : self.charge_history,
+                    'details' : self.details.serialize()
+                }
+
+    def deserialize(self, data):
+        try:
+            self.user_id = data['user_id']
+            self.nickname = data['nickname']
+            self.payment_type = data['payment_type']
+            self.is_default = False
+            self.is_removed = False
+            self.charge_history = 0.0
+            details = data['details']
+            d = Detail()
+            if (self.payment_type == 'credit' or self.payment_type == 'debit'):
+                self.details = d.deserialize_card(details)
+            else:
+                self.details = d.deserialize_paypal(details)
+        except KeyError as e:
+            raise DataValidationError('Invalid payment: missing ' + e.args[0])
+        except TypeError as e:
+            raise DataValidationError('Invalid payment: body of request contained bad or no data')
+        return self
```

Comment:
I.... don't know. An artifact from starting with the lecture materials, which do the same. But looking back at the TDD lecture code, i don't know why Rofrano's deserialize returns self either. Either way, striking the returns.